### ISSUE TITLE
Unneeded packages end up in production deployment

### ISSUE THREAD

**Issue by janconc** [prisma/prisma#17720]
### Bug description

Motivation:  I use Prisma in an Azure function app. It's starting time depends on the app size.

Situation: my package.json contains (among others):
```json
{
  "dependencies": {
    "@prisma/client": "^4.9.0"
  },
  "devDependencies": {
    "prisma": "^4.9.0",
  }
}
```
After packages installing, app building and pruning I end up with packages that are not needed (?) to run the application, but take a lot of space
```
du -h --max-depth=1 node_modules/@prisma | sort -h
32K     node_modules/@prisma/engines-version
2.9M    node_modules/@prisma/client
59M     node_modules/@prisma/engines
62M     node_modules/@prisma
```
It looks like the @prisma/engine is not needed but is not removed by pruning command, i.e. `npm prune --production` or `yarn install --production`

I am not an expert in node packaging but If I understood it correctly, only packages listed in `devDependencies` are removed by above pruning commands.

Or what is the correct way to do this?

### How to reproduce

1. Have node.js prisma app
2. Install packages: `yarn install`
3. Build your app, i.e. `yarn build`
4. Remove development time packages: `npm prune --production`
5. check whether only the packages needed to run the app are present in node_modules
Result: There is a 60 MB package node_modules/@prisma/engines

### Expected behavior

only the packages needed to run the app are present in node_modules. In practice @prisma/engine should not be there

### Prisma information

version 4.9

### Environment & setup

Windows + WSL2 + Ubuntu image, happens probably in all environments.

### Prisma Version

```
prisma                  : 4.9.0
@prisma/client          : 4.9.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Format Wasm             : @prisma/prisma-fmt-wasm 4.9.0-42.ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5
Default Engines Hash    : ceb5c99003b99c9ee2c1d2e618e359c14aef2ea5
Studio                  : 0.479.0
```

**Comment by janpio** [comment]
What version of yarn are you using to install your dependencies? Yarn 1 and 2+ are _very_ different-

**Comment by janpio** [comment]
We looked into this a bit @janconc. Our current theory is as follows:

When you install `@prisma/client`, it also defines `prisma` as an peer dependency. 
This means that in the package tree, `prisma` is seen as a child of `@prisma/client` (via `npm list --depth 2`):

```
17720@1.0.0 C:\Users\Jan\Documents\throwaway\17720
├─┬ @prisma/client@4.10.0
│ ├── @prisma/engines-version@4.10.0-84.ca7fcef713137fa11029d519a9780db130cca91d
│ └── prisma@4.10.0 deduped
└─┬ prisma@4.10.0
  └── @prisma/engines@4.10.0
```

As you can see `@prisma/engines`, the big thing from your `du -h` command, is a child of `prisma` CLI.
Now when you run `npm prune --production`, it unfortunately does _not_ prune `prisma` although it is just a peer dependency. It is even marked as `optional:true` in the `peerDependenciesMeta`:
```
  "peerDependencies": {
    "prisma": "*"
  },
  "peerDependenciesMeta": {
    "prisma": {
      "optional": true
    }
  },
```

We are not certain that this is really intended that `npm prune` keeps this around, and I will have to ask internally if someone knows more about the topic.

**Comment by yusufkandemir** [comment]
The issues are not related, but as can be seen in https://github.com/prisma/prisma/issues/15852#issuecomment-1423251790, `pnpm deploy --prod` also brings `@prisma/engines`.

The relevant part of `pnpm list --depth 2`:
```
dependencies:
@prisma/client 4.10.0
├── @prisma/engines-version 4.10.0-84.ca7fcef713137fa11029d519a9780db130cca91d
└─┬ prisma 4.10.0 peer
  └── @prisma/engines 4.10.0
```

**Comment by isaachinman** [comment]
@janpio Can you comment if it's safe to manually delete the `node_modules/prisma` dir in a production release process?

**Comment by janpio** [comment]
Yes, the `prisma` CLI package is not needed at all by Prisma Client. All the files should be contained in `@prisma/client` and `.prisma` - and if you use a custom `output` folder even just in that folder.

**Comment by tarunregmi** [comment]
I tried this approach (here: https://github.com/prisma/prisma/discussions/24910) (`output` folder).

But there's a problem with this approach, `prisma migrate` doesn't work here in `production` mode.

Or is there any other way to perform `prisma migrate` in this approach?

**Comment by tarunregmi** [comment]
> Yes, the `prisma` CLI package is not needed at all by Prisma Client. All the files should be contained in `@prisma/client` and `.prisma` - and if you use a custom `output` folder even just in that folder.

@janpio 

I tried this approach (here: https://github.com/prisma/prisma/discussions/24910) (`output` folder).

But there's a problem with this approach, `prisma migrate` doesn't work here in `production` mode.

Or are there any other way to perform `prisma migrate` in this approach?

**Comment by tarunregmi** [comment]
Here is my solution:

Install `prisma-client` as a `devDependency` and use a custom folder (`output`) for `prisma-client`. This way, you can remove all unnecessary dependencies in production using `npm prune --production`.

This solution works in most cases in production because, in many scenarios, we use `prisma-client` only in production. However, there is a problem with this approach: you will not be able to run `prisma-cli` commands (like `prisma migrate`, etc.) in production.

There is a solution for this issue as well:
1. Install all dependencies: `npm i`
2. Run the required `prisma` commands: `prisma migrate`, etc.
3. Remove unnecessary dependencies: `npm prune --production`

**Comment by baraknaveh** [comment]
**Workaround: `.pnpmfile.cjs`**

If you're using pnpm, you can use a [pnpm hook](https://pnpm.io/pnpmfile) to strip the problematic optional peer at resolution time:

```js
function readPackage(pkg) {
  if (pkg.name === '@prisma/client') {
    delete pkg.peerDependencies?.prisma
    delete pkg.peerDependenciesMeta?.prisma
  }
  return pkg
}

module.exports = { hooks: { readPackage } }
```

This prevents `prisma` (and its transitive tree) from being resolved under `@prisma/client`, so `pnpm deploy --prod` stays clean. `prisma generate` still works — the CLI is still in devDependencies, it's just no longer linked as a runtime peer.

For the curious: `ignoreMissing`, `auto-install-peers=false`, and moving the dep to root all failed — the lockfile is surprisingly sticky about resolved optional peers.
